File: benchmarks/producer.py
#!/usr/bin/env python3
"""
Producer for benchmarking: sends random bytes via STDIO or shared memory.
"""

import argparse
import os
import random
import struct
import sys
import time
from typing import Optional

# Add parent directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'python'))

from ipc0cp.ring_buffer import SharedRingBufferProducer


def produce_stdio(min_size: int, max_size: int, duration: float) -> dict:
    """
    Produce random bytes via STDOUT (binary mode).
    
    Protocol: [length:8 bytes][payload:length bytes]
    
    Args:
        min_size: Minimum payload size in bytes
        max_size: Maximum payload size in bytes
        duration: How long to run in seconds
        
    Returns:
        Statistics dict with bytes_sent, messages_sent, elapsed_time
    """
    # Use binary stdout
    stdout = sys.stdout.buffer
    
    bytes_sent = 0
    messages_sent = 0
    start_time = time.time()
    
    while True:
        elapsed = time.time() - start_time
        if elapsed >= duration:
            break
        
        # Generate random payload
        size = random.randint(min_size, max_size)
        payload = os.urandom(size)
        
        # Send length prefix (8 bytes, little-endian uint64)
        length_bytes = struct.pack('<Q', size)
        stdout.write(length_bytes)
        
        # Send payload
        stdout.write(payload)
        stdout.flush()
        
        bytes_sent += size
        messages_sent += 1
    
    # Send end-of-stream marker (length=0)
    end_marker = struct.pack('<Q', 0)
    stdout.write(end_marker)
    stdout.flush()
    
    elapsed_time = time.time() - start_time
    
    return {
        'bytes_sent': bytes_sent,
        'messages_sent': messages_sent,
        'elapsed_time': elapsed_time,
        'throughput_mbps': (bytes_sent / elapsed_time) / (1024 * 1024),
    }


def produce_shm(shm_name: str, min_size: int, max_size: int, duration: float) -> dict:
    """
    Produce random bytes via shared memory ring buffer.
    
    Args:
        shm_name: Shared memory segment name
        min_size: Minimum payload size in bytes
        max_size: Maximum payload size in bytes
        duration: How long to run in seconds
        
    Returns:
        Statistics dict with bytes_sent, messages_sent, elapsed_time
    """
    # Create producer with 2GB buffer
    producer = SharedRingBufferProducer(
        shm_name=shm_name,
        total_data_bytes=2 * 1024 * 1024 * 1024,  # 2 GB
        blocking=True,
        max_slot_size=10 * 1024 * 1024,  # 10 MB
    )
    
    bytes_sent = 0
    messages_sent = 0
    start_time = time.time()
    
    try:
        while True:
            elapsed = time.time() - start_time
            if elapsed >= duration:
                break
            
            # Generate random payload
            size = random.randint(min_size, max_size)
            payload = os.urandom(size)
            
            # Push to ring buffer
            success = producer.push(payload, timeout=5.0)
            
            if success:
                bytes_sent += size
                messages_sent += 1
            else:
                # Timeout - consumer might be slow
                pass
        
        elapsed_time = time.time() - start_time
        
        return {
            'bytes_sent': bytes_sent,
            'messages_sent': messages_sent,
            'elapsed_time': elapsed_time,
            'throughput_mbps': (bytes_sent / elapsed_time) / (1024 * 1024),
        }
    finally:
        # Close sends EOS marker and cleans up
        producer.close()
        producer.unlink()


def main():
    parser = argparse.ArgumentParser(description='Benchmark producer')
    
    # Transport mode
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument('--stdio', action='store_true', help='Use STDIN/STDOUT')
    group.add_argument('--shm', type=str, help='Shared memory segment name')
    
    # Benchmark parameters
    parser.add_argument('--min-size', type=int, default=512 * 1024, 
                        help='Minimum payload size in bytes (default: 512KB)')
    parser.add_argument('--max-size', type=int, default=5 * 1024 * 1024,
                        help='Maximum payload size in bytes (default: 5MB)')
    parser.add_argument('--duration', type=float, default=60.0,
                        help='Duration to run in seconds (default: 60)')
    parser.add_argument('--quiet', action='store_true',
                        help='Suppress progress output')
    
    args = parser.parse_args()
    
    if not args.quiet:
        print(f"Producer starting: min_size={args.min_size}, max_size={args.max_size}, duration={args.duration}",
              file=sys.stderr)
    
    if args.stdio:
        stats = produce_stdio(args.min_size, args.max_size, args.duration)
    else:
        stats = produce_shm(args.shm, args.min_size, args.max_size, args.duration)
    
    # Print stats to stderr so it doesn't interfere with STDIO mode
    print(f"\nProducer Stats:", file=sys.stderr)
    print(f"  Bytes sent: {stats['bytes_sent']:,} ({stats['bytes_sent'] / (1024**3):.2f} GB)", file=sys.stderr)
    print(f"  Messages sent: {stats['messages_sent']:,}", file=sys.stderr)
    print(f"  Elapsed time: {stats['elapsed_time']:.2f} seconds", file=sys.stderr)
    print(f"  Throughput: {stats['throughput_mbps']:.2f} MB/s", file=sys.stderr)


if __name__ == '__main__':
    main()

File: benchmarks/producer.cpp
#include "ipc0cp/ring_buffer.hpp"
#include "ipc0cp/stdio.hpp"

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <cstring>
#include <iostream>
#include <random>
#include <string>
#include <string_view>
#include <vector>

using namespace ipc0cp;

namespace {

struct Config {
    bool use_stdio = false;
    bool use_shm = false;
    std::string shm_name;

    double duration = 60.0;
    size_t min_size = 512 * 1024;
    size_t max_size = 5 * 1024 * 1024;
    bool quiet = false;

    size_t shm_total_data_bytes = 2ULL * 1024 * 1024 * 1024;
};

bool parse_args(int argc, char** argv, Config& config) {
    for (int i = 1; i < argc; ++i) {
        std::string_view arg(argv[i]);
        if (arg == "--stdio") {
            config.use_stdio = true;
        } else if (arg == "--shm" && i + 1 < argc) {
            config.use_shm = true;
            config.shm_name = std::string(argv[++i]);
        } else if (arg == "--duration" && i + 1 < argc) {
            config.duration = std::stod(argv[++i]);
        } else if (arg == "--min-size" && i + 1 < argc) {
            config.min_size = std::stoull(argv[++i]);
        } else if (arg == "--max-size" && i + 1 < argc) {
            config.max_size = std::stoull(argv[++i]);
        } else if (arg == "--quiet") {
            config.quiet = true;
        }
    }

    if (config.min_size > config.max_size) {
        std::swap(config.min_size, config.max_size);
    }

    if (config.use_stdio == config.use_shm) {
        std::cerr << "Exactly one of --stdio or --shm <name> is required" << std::endl;
        return false;
    }

    if (config.use_shm && config.shm_name.empty()) {
        std::cerr << "Missing shm name for --shm" << std::endl;
        return false;
    }

    return true;
}

void fill_random(std::vector<uint8_t>& payload, std::mt19937_64& rng) {
    size_t filled = 0;
    const size_t payload_size = payload.size();
    while (filled < payload_size) {
        uint64_t value = rng();
        size_t copy = std::min(payload_size - filled, sizeof(value));
        std::memcpy(payload.data() + filled, &value, copy);
        filled += copy;
    }
}

}  // namespace

int main(int argc, char** argv) {
    Config config;
    if (!parse_args(argc, argv, config)) {
        return 2;
    }

    std::mt19937_64 rng(std::random_device{}());
    std::uniform_int_distribution<size_t> size_dist(config.min_size, config.max_size);

    const std::string metadata = R"({"type":"bytes"})";
    std::vector<uint8_t> payload;
    payload.reserve(config.max_size);

    size_t bytes_sent = 0;
    size_t messages = 0;
    auto start = std::chrono::steady_clock::now();

    if (config.use_stdio) {
        StdioProducer producer;

        while (true) {
            auto elapsed = std::chrono::duration<double>(
                std::chrono::steady_clock::now() - start
            ).count();
            if (elapsed >= config.duration) {
                break;
            }

            const size_t payload_size = size_dist(rng);
            payload.resize(payload_size);
            fill_random(payload, rng);

            producer.push_raw(metadata, payload);
            bytes_sent += payload_size;
            ++messages;
        }

        producer.close();
    } else {
        SharedRingBufferProducer producer(
            config.shm_name,
            config.shm_total_data_bytes,
            true
        );

        while (true) {
            auto elapsed = std::chrono::duration<double>(
                std::chrono::steady_clock::now() - start
            ).count();
            if (elapsed >= config.duration) {
                break;
            }

            const size_t payload_size = size_dist(rng);
            payload.resize(payload_size);
            fill_random(payload, rng);

            if (producer.push_raw(metadata, payload, 5000)) {
                bytes_sent += payload_size;
                ++messages;
            }
        }

        producer.close();
    }

    const auto elapsed = std::chrono::duration<double>(
        std::chrono::steady_clock::now() - start
    ).count();

    const double throughput = elapsed > 0
        ? (static_cast<double>(bytes_sent) / (1024.0 * 1024.0)) / elapsed
        : 0.0;

    std::cerr << "\nProducer Stats:\n";
    std::cerr << "  Bytes sent (payload only): " << bytes_sent << "\n";
    std::cerr << "  Messages sent: " << messages << "\n";
    std::cerr << "  Elapsed time: " << elapsed << " seconds\n";
    std::cerr << "  Throughput: " << throughput << " MB/s\n";

    return 0;
}
